### PR TITLE
Add support for precertificate verification.

### PR DESCRIPTION
**Pull request by plietar**
In addition to the already supported plain X.509 certificate,
precertificates and certificates embedding an SCT list can be verified.
PreCertificates signed with special purpose CA certificates are also
supported.

### CODE REVIEW

File: python/ct/crypto/asn1/oid.py [hunk 1]
```diff
@@ -327,7 +332,11 @@ def _decode_value(cls, buf, strict=True):
     ID_AD_CA_ISSUERS: ("id-ad-caIssuers", "caIssuers"),
     ID_QT_CPS: ("id-qt-cps", "CPS"),
     ID_QT_UNOTICE: ("id-qt-unotice", "UserNotice"),
-    ANY_POLICY: ("anyPolicy", "anyPolicy")
+    ANY_POLICY: ("anyPolicy", "anyPolicy"),
+
+    EMBEDDED_SCT_LIST: ("ctEmbeddedSCT", "ctEmbeddedSCT"),
+    POISON: ("ctPoison", "ctPoison"),
+    PRECERTIFICATE_SIGNING: ("ctPresign", "ctPresign")
```
Comment (eranmes) [inline python/ct/crypto/asn1/oid.py]: Nit: The more informative name is Precertificate Signing Cert

File: python/ct/crypto/tools/verify_util.py [hunk 2]
```diff
@@ -13,7 +13,7 @@
 
   For example:
 
-  verify_util.py verify_sct --sct=cert_sct.tls --log_key=log_key.pem cert.pem
+  verify_util.py verify_sct --sct=cert_sct.tls --log_key=log_key.pem cert.pem [cert.pem...]
```
Comment (eranmes) [inline python/ct/crypto/tools/verify_util.py]: This chain should not be necessary - a single PEM file can contain multiple certificates and that's the convention around the codebase.

File: python/ct/crypto/tools/verify_util.py [hunk 3]
```diff
@@ -61,15 +61,20 @@ def main(argv):
...
     if command != "verify_sct":
         exit_with_message("Unknown command %s" % command)
 
-    if not cert_file:
+    if not chain:
         exit_with_message("No certificate file given")
 
-    verify_sct(cert.Certificate.from_pem_file(cert_file, strict_der=False),
+    chain = map(lambda name:
+                    cert.Certificate.from_pem_file(name, strict_der = True),
```
Comment (eranmes) [inline python/ct/crypto/tools/verify_util.py]: The Certificate.from_pem_file returns an iterable of certs from the file, which is why you don't need to specify multiple files.

File: python/ct/crypto/cert.py [hunk 4]
```diff
@@ -184,11 +184,15 @@ def to_der(self):
         return self._asn1_cert.encode()
 
     def to_pem(self):
-      return pem.to_pem(self._asn1_cert.encode(), self.PEM_MARKERS[0])
+        return pem.to_pem(self._asn1_cert.encode(), self.PEM_MARKERS[0])
```
Comment (eranmes) [inline python/ct/crypto/cert.py]: Unnecessary whitespace change.

File: python/ct/crypto/verify.py [hunk 5]
```diff
@@ -211,24 +333,15 @@ def verify_sct(self, sct, chain):
             ct.crypto.error.EncodingError: failed to encode signature input,
                 or decode the signature.
             ct.crypto.error.SignatureError: invalid signature.
+            ct.crypto.error.MissingCertificateError: a certificate is missing.
```
Comment (eranmes) [inline python/ct/crypto/verify.py]: nit: Add "from the chain".
A better error would be "MissingIssuerCertificate" so it's clear which one is missing (and why).

File: python/ct/crypto/verify.py [hunk 6]
```diff
@@ -50,6 +53,128 @@ def decode_signature(signature):
...
+            })
+        ]
+
+    return asn1["tbsCertificate"].encode()
+
+def is_precertificate(cert):
+    return (cert.has_extension(oid.CT_POISON) or
+            cert.has_extension(oid.CT_EMBEDDED_SCT_LIST))
+
+def create_dst_entry(sct, chain):
```
Comment (eranmes) [inline python/ct/crypto/verify.py]: Make this a private method.

File: python/ct/crypto/verify.py (hunk 5 above)
Comment (plietar) [inline python/ct/crypto/verify.py]: The error is also used when the leaf certificate is missing

Comment (eranmes) [inline python/ct/crypto/verify.py]: In this case, how about IncompleteChainError? that would be more indicative of the error.

File: python/ct/crypto/cert.py [hunk 7]
```diff
@@ -378,21 +382,33 @@ def subject_alternative_names(self):
         sans = self._get_decoded_extension_values(oid.ID_CE_SUBJECT_ALT_NAME)
         return sum([list(san) for san in sans], [])
 
-    def subject(self):
-        """Returns list of subject field values in fashion similar to
-        subject_alternative_names fashion.
+    def subject(self, flatten=True):
```
Comment (eranmes) [inline python/ct/crypto/cert.py]: The change here is for a very specific use case so I recommend leaving it out of the general API.
Instead, get the value from the asn1 the in verify.py - the same way you do for the other extensions.

File: python/ct/crypto/verify.py [hunk 8]
```diff
@@ -50,6 +53,129 @@ def decode_signature(signature):
                                    len(remaining)))
     return (hash_algo, sig_algo, remaining)
 
+def get_issuer(chain):
```
Comment (eranmes) [inline python/ct/crypto/verify.py]: prefix with underscore to indicate private method and name _get_precertificate_issuer so it's clear this is useful in precertificate serialization context.

File: python/ct/crypto/verify.py [hunk 9]
```diff
@@ -50,6 +53,129 @@ def decode_signature(signature):
...
+        return issuer
+    else:
+        try:
+            return chain[2]
+        except IndexError:
+            raise error.MissingCertificateError(
+                "Chain with PreCertificate signed by PreCertificate "
+                "Signing Cert must contain issuer.")
+
+def find_extension(asn1, extn_id):
```
Comment (eranmes) [inline python/ct/crypto/verify.py]: prefix with underscore to indicate it shouldn't be used outside this module.

File: python/ct/crypto/verify.py [hunk 10]
```diff
@@ -50,6 +53,129 @@ def decode_signature(signature):
...
+        The decoded value of the extension, or None if not found.
+        This is a reference and can be modified.
+    """
+    for e in asn1["tbsCertificate"]["extensions"]:
+        if e["extnID"] == extn_id:
+            return e["extnValue"].decoded_value
+
+    return None
+
+def remove_extension(asn1, extn_id):
```
Comment (eranmes) [inline python/ct/crypto/verify.py]: Same as for find_extension.

File: python/ct/crypto/verify.py [hunk 11]
```diff
@@ -50,6 +53,129 @@ def decode_signature(signature):
...
+    Args:
+        asn1: x509.Certificate instance.
+        extn_id: OID of the extension to be removed.
+    """
+    asn1["tbsCertificate"]["extensions"] = (
+        filter(lambda e: e["extnID"] != extn_id,
+               asn1["tbsCertificate"]["extensions"])
+    )
+
+def encode_tbs_certificate_for_validation(cert, issuer):
```
Comment (eranmes) [inline python/ct/crypto/verify.py]: Same.

**Comment by eranmes** [comment]
(hopefully) final round of review comments - should be good to go once those are addressed.

File: python/ct/crypto/verify_test.py [hunk 12]
```diff
@@ -249,37 +249,71 @@ def test_verify_sth_consistency_invalid_proof(self):
...
-    def test_verify_sct_valid_signature(self):
-        root_cert = cert.Certificate.from_pem_file(
-            os.path.join(FLAGS.testdata_dir, 'ca-cert.pem'))
-        test_cert = cert.Certificate.from_pem_file(
-            os.path.join(FLAGS.testdata_dir, 'test-cert.pem'))
+    def _test_verify_sct(self, proof, chain, timestamp = None):
         sct = client_pb2.SignedCertificateTimestamp()
-        tls_message.decode(read_testdata_file('test-cert.proof'), sct)
+        tls_message.decode(read_testdata_file(proof), sct)
+        if timestamp is not None:
```
Comment (eranmes) [inline python/ct/crypto/verify_test.py]: In which case would the timestamp be None? you should have it for all test data, don't you?

Comment (plietar) [inline python/ct/crypto/verify_test.py]: `timestamp` is a fake value used to modify the SCT to make timestamp validation fail.
The real value is loaded just above.

Comment (plietar) [inline python/ct/crypto/verify_test.py]: I'll change the name to make it more obvious
